x/qgb/README.md

```diff
@@ -0,0 +1,141 @@
+
+I.e. the current block height is not 0, and we're at a data commitment window height.
+
+The data commitment window is defined as a governance parameter that can be changed. Currently, it is set to 400.
+
+### Panics
+
+During EndBlock step, the state machine generates new attestations if needed. During this generation, the state machine could panic.
+
+#### Data commitment panics
+
+During EndBlock, if the block height corresponds to a `DataCommitmentWindow`, it will generate a new data commitment, during which, the state machine can panic in the following case:
+
+- An unexpected behavior happened while getting the current data commitment:
+
+```golang
+dataCommitment, err := k.GetCurrentDataCommitment(ctx)  
+if err != nil {  
+   panic(sdkerrors.Wrap(err, "coudln't get current data commitment"))  
+}
+```
+
+#### Valset panics
+
+Similar to data commitments, when checking if the state machine needs to generate a new valset, it might panic in the following cases:
+
+- When checking that a previous valset has been emitted, but it is unable to get it:
+
+```golang
+if k.CheckLatestAttestationNonce(ctx) && k.GetLatestAttestationNonce(ctx) != 0 {  
+   var err error  
+   latestValset, err = k.GetLatestValset(ctx)  
+   if err != nil {  
+      panic(err)  
+   }  
+}
+```
+
+- When getting the current valset:
+
+```golang
+vs, err := k.GetCurrentValset(ctx)
+if err != nil {  
+   // this condition should only occur in the simulator  
+   // ref : https://github.com/Gravity-Bridge/Gravity-Bridge/issues/35 
+   if errors.Is(err, types.ErrNoValidators) {  
+      ctx.Logger().Error("no bonded validators",  
+         "cause", err.Error(),  
+      )  
+      return  
+   }  
+   panic(err)  
+}
+```
+
+- When creating the internal validator struct, i.e. mapping the validators EVM addresses to their powers:
+
+```golang
+intLatestMembers, err := types.BridgeValidators(latestValset.Members).ToInternal()  
+if err != nil {  
+   panic(sdkerrors.Wrap(err, "invalid latest valset members"))  
+}
+```
+
+#### Attestations panics
+
+When storing a new attestation, which is either a data commitment or a valset, the state machine can panic in the following cases:
+
+- The attestation request created from the data commitment is a duplicate of an existing attestation:
+
+```golang
+key := []byte(types.GetAttestationKey(nonce))  
+store := ctx.KVStore(k.storeKey)  
+  
+if store.Has(key) {  
+   panic("trying to overwrite existing attestation request")  
+}
+```
+
+- An error happened while marshalling the interface:
+
+```golang
+b, err := k.cdc.MarshalInterface(at)  
+if err != nil {  
+   panic(err)  
+}
+```
+
+- The universal nonce was not incremented correctly by 1:
+
+```golang
+if k.CheckLatestAttestationNonce(ctx) && k.GetLatestAttestationNonce(ctx)+1 != nonce {  
+   panic("not incrementing latest attestation nonce correctly")  
+}
+```
+
+## Useful links
```

Comment:
could make link the ADRs

Reply:
[13e6376](https://github.com/celestiaorg/celestia-app/pull/1052/commits/13e6376199508e32ec8c5524c7d96869b69a028e)